File: src/jarvis/api/routes/mcp.py
"""MCP server management (S4, ADR 0012 §5). Servers are tenant-scoped rows:
CRUD is admin/owner only (anonymous mode is the default tenant's full
access), listing/probing is any tenant member, and a foreign tenant's
server id is 404 — no existence leak (D29). Names are immutable once
created (they join agent version snapshots via `mcp__<name>__<tool>`), so
a delete leaves agents' snapshots intact — the next run fails resolution
honestly with a terminal `tool` failure (D38)."""

from __future__ import annotations

from uuid import uuid4

from fastapi import APIRouter, Depends
from pydantic import ValidationError
from sqlalchemy.exc import IntegrityError

from jarvis.api.auth import AuthContext, AuthDep
from jarvis.api.deps import AppContainer, get_container
from jarvis.api.errors import ApiError
from jarvis.api.schemas import McpProbeResponse, McpServerCreate, McpServerList, McpServerPatch
from jarvis.domain.mcp import McpServer
from jarvis.tools.mcp.errors import McpResolutionError

router = APIRouter(prefix="/mcp", tags=["mcp"])

# Module-level Depends singleton (ruff B008).
ContainerDep = Depends(get_container)


def _require_server_manager(auth: AuthContext) -> None:
    """Admin/owner only — except anonymous mode, the default tenant's full
    access (the dev walkthrough needs no auth setup)."""
    if auth.user is not None and auth.principal.role not in ("owner", "admin"):
        raise ApiError(403, "forbidden", "MCP server management requires the admin or owner role")


async def _visible_server(auth: AuthContext, container: AppContainer, server_id: str) -> McpServer:
    server = await container.mcp_servers.get(server_id, tenant_id=auth.principal.tenant_id)
    if server is None:
        raise ApiError(404, "not_found", f"MCP server {server_id!r} not found")
    return server


@router.get("/servers")
async def list_servers(
    auth: AuthContext = AuthDep, container: AppContainer = ContainerDep
) -> McpServerList:
    servers = await container.mcp_servers.list_servers(tenant_id=auth.principal.tenant_id)
    return McpServerList(items=servers)


@router.post("/servers", status_code=201)
async def create_server(
    req: McpServerCreate, auth: AuthContext = AuthDep, container: AppContainer = ContainerDep
) -> McpServer:
    _require_server_manager(auth)
    try:
        server = McpServer(id=str(uuid4()), name=req.name, config=req.config, enabled=req.enabled)
    except ValidationError as exc:
        raise ApiError(422, "validation", exc.errors()[0]["msg"]) from None
    try:
        return await container.mcp_servers.create(server, tenant_id=auth.principal.tenant_id)
    except IntegrityError:
        raise ApiError(
            409, "conflict", f"an MCP server named {req.name!r} already exists"
        ) from None


@router.get("/servers/{server_id}")
async def get_server(
    server_id: str, auth: AuthContext = AuthDep, container: AppContainer = ContainerDep
) -> McpServer:
    return await _visible_server(auth, container, server_id)


@router.patch("/servers/{server_id}")
async def update_server(
    server_id: str,
    req: McpServerPatch,
    auth: AuthContext = AuthDep,
    container: AppContainer = ContainerDep,
) -> McpServer:
    _require_server_manager(auth)
    if req.name is not None:
        raise ApiError(
            422,
            "validation",
            "MCP server names are immutable — they join agent version snapshots "
            "(delete the server and create a new one instead)",
        )
    current = await _visible_server(auth, container, server_id)
    updated = current.model_copy(
        update={
            "enabled": req.enabled if req.enabled is not None else current.enabled,
            "config": req.config if req.config is not None else current.config,
        }
    )
    try:
        return await container.mcp_servers.update(updated, tenant_id=auth.principal.tenant_id)
    except LookupError:
        raise ApiError(404, "not_found", f"MCP server {server_id!r} not found") from None


@router.delete("/servers/{server_id}", status_code=204)
async def delete_server(
    server_id: str, auth: AuthContext = AuthDep, container: AppContainer = ContainerDep
) -> None:
    _require_server_manager(auth)
    deleted = await container.mcp_servers.delete(server_id, tenant_id=auth.principal.tenant_id)
    if not deleted:
        raise ApiError(404, "not_found", f"MCP server {server_id!r} not found")


@router.post("/servers/{server_id}/probe")
async def probe_server(
    server_id: str, auth: AuthContext = AuthDep, container: AppContainer = ContainerDep
) -> McpProbeResponse:
    """Connect fresh through the provider's connection path — no registry
    side effects, nothing persisted. Any member may probe; a resolution
    failure is a 502 in the frozen error envelope (the server is reachable
    enough to exist but not to talk to)."""
    server = await _visible_server(auth, container, server_id)
    try:
        # The requesting principal rides along: stored header refs (ADR
        # 0013) decrypt tenant-scoped for whoever probes — a member probing
        # the same server gets the same behavior as an admin.
        tools = await container.mcp.probe(server, principal=auth.principal)
    except McpResolutionError as exc:
        raise ApiError(502, "mcp_unreachable", str(exc)) from None
    return McpProbeResponse(server=server, tools=tools)


__all__ = ["router"]

File: web/src/api/queries/mcp.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { client, unwrap } from "@/api/client";
import { apiErrorFromResponse } from "@/api/errors";

import type { components } from "@/api/schema";

// MCP server management (S4, ADR 0012 §5): tenant-scoped rows, admin/owner
// writes (anonymous mode = full access), any-member list/probe. Types come
// from the generated schema; the 204 delete goes through the raw client.

export type McpServer = components["schemas"]["McpServer"];
export type McpServerCreate = components["schemas"]["McpServerCreate"];
export type McpServerPatch = components["schemas"]["McpServerPatch"];
export type McpProbeResponse = components["schemas"]["McpProbeResponse"];

// ADR 0013: header/env refs widen to the stored variant — the same
// credential-reference union the model credential_ref uses.
export type CredentialRef =
  | components["schemas"]["EnvCredentialRef"]
  | components["schemas"]["StoredCredentialRef"];

export const mcpServersQueryKey = ["mcp-servers"] as const;

export function useMcpServers() {
  return useQuery({
    queryKey: mcpServersQueryKey,
    queryFn: async () => {
      const body = await unwrap(client.GET("/v1/mcp/servers"));
      return body.items;
    },
  });
}

export function useCreateMcpServer() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (body: McpServerCreate) =>
      unwrap(client.POST("/v1/mcp/servers", { body })),
    onSuccess: () => void queryClient.invalidateQueries({ queryKey: mcpServersQueryKey }),
  });
}

export function useUpdateMcpServer() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ serverId, body }: { serverId: string; body: McpServerPatch }) =>
      unwrap(client.PATCH("/v1/mcp/servers/{server_id}", {
        params: { path: { server_id: serverId } },
        body,
      })),
    onSuccess: () => void queryClient.invalidateQueries({ queryKey: mcpServersQueryKey }),
  });
}

export function useDeleteMcpServer() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (serverId: string) => {
      const { response, error } = await client.DELETE("/v1/mcp/servers/{server_id}", {
        params: { path: { server_id: serverId } },
      });
      if (!response.ok) throw apiErrorFromResponse(response, error);
    },
    onSuccess: () => void queryClient.invalidateQueries({ queryKey: mcpServersQueryKey }),
  });
}

// Probe is a POST (it connects fresh and persists nothing), but it is a
// read semantically — a query keyed by server id keeps the "View tools"
// interaction declarative; a failed connect surfaces as the 502 envelope.
export function useProbeMcpServer(serverId: string) {
  return useQuery({
    queryKey: ["mcp-probe", serverId],
    enabled: serverId !== "",
    queryFn: async () => {
      const body = await unwrap(client.POST("/v1/mcp/servers/{server_id}/probe", {
        params: { path: { server_id: serverId } },
      }));
      return body;
    },
  });
}